src/Compilers/Core/Portable/Syntax/SyntaxTree.cs

```diff
@@ -399,5 +406,20 @@ internal virtual bool SupportsLocations
         {
             get { return this.HasCompilationUnitRoot; }
         }
+
+        internal SourceGeneratorSyntaxTreeInfo GetSourceGeneratorInfo(
```

Comment:
Should `SourceGeneratorSyntaxTreeInfo` be changed to a value type?

If we did that, I anticipated that we would just store the flags we want to track in some easy-to-synchronize form and then return instances of the struct on-demand.

Reply:
I wasn't sure the best way to atomically update a nullable value type :D   I can change to that, but would need info on how best to handle that part of hte problem so we don't have to deal with struct tearing.

Reply:
let me tweak the rest of this though.

Reply:
changed this.  lmk if it's better.